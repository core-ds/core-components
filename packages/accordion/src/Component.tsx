import React, {
    AnchorHTMLAttributes,
    FC,
    KeyboardEvent,
    ReactNode,
    useCallback,
    useState,
} from 'react';
import cn from 'classnames';

import { Typography } from '@alfalab/core-components-typography';

import { DefaultControlIcon } from './components';
import { useRecalculateContentHeight } from './hooks';
import { ControlPosition } from './typings';

import styles from './index.module.css';

export type AccordionProps = {
    /**
     * Состояние компонента
     */
    expanded?: boolean;

    /**
     * Элемент заголовка
     */
    header: ReactNode;

    /**
     * Слот для элемента управления
     */
    control?: ReactNode;

    /**
     * Указывает компоненту - где будет размещен control
     */
    controlPosition?: ControlPosition;

    /**
     * Начальное состояние uncontrolled компонента
     */
    defaultExpanded?: boolean;

    /**
     * Основной элемент для отображения содержимого
     */
    children?: ReactNode;

    /**
     * Дополнительный класс обертки
     */
    className?: string;

    /**
     * Дополнительный класс для контейнера с заголовком
     */
    containerClassName?: string;

    /**
     * Дополнительный класс для header
     */
    headerClassName?: string;

    /**
     * Дополнительный класс для control
     */
    controlClassName?: string;

    /**
     * Дополнительный класс для body
     */
    bodyClassName?: string;

    /**
     * Обработчик смены состояний `expanded`
     */
    onExpandedChange?: (expanded: boolean) => void;

    /**
     * Идентификатор для систем автоматизированного тестирования
     */
    dataTestId?: string;
} & AnchorHTMLAttributes<HTMLDivElement>;

export const Accordion: FC<AccordionProps> = ({
    expanded,
    defaultExpanded = false,
    header,
    control,
    controlPosition = 'end',
    children,
    className,
    containerClassName,
    headerClassName,
    controlClassName,
    bodyClassName,
    onExpandedChange,
    dataTestId,
    ...rest
}) => {
    const uncontrolled = expanded === undefined;
    const [expandedState, setExpanded] = useState(uncontrolled ? defaultExpanded : expanded);
    const isExpanded = uncontrolled ? expandedState : expanded;

    const isStartPosition = controlPosition === 'start';

    const { contentRef, contentCaseRef } = useRecalculateContentHeight(isExpanded);

    const controlContent =
        control === undefined ? (
            <DefaultControlIcon expanded={isExpanded} startPosition={isStartPosition} />
        ) : (
            control
        );

    const headerContent =
        typeof header === 'string' ? (
            <Typography.Text view='primary-large' weight='medium'>
                {header}
            </Typography.Text>
        ) : (
            header
        );

    const bodyContent =
        typeof children === 'string' ? (
            <Typography.Text view='primary-medium'>{children}</Typography.Text>
        ) : (
            children
        );

    const handleExpandedChange = useCallback(() => {
        if (uncontrolled) {
            setExpanded(!isExpanded);
        }

        onExpandedChange?.(!isExpanded);
    }, [isExpanded, onExpandedChange, uncontrolled]);

    const handleKeyDown = (event: KeyboardEvent) => {
        if (event.key === 'Enter') {
            handleExpandedChange();
        }
    };

    return (
        <div {...rest} data-test-id={dataTestId} className={cn(styles.accordion, className)}>
            <div
                role='button'
                tabIndex={0}
                aria-expanded={isExpanded}
                onClick={handleExpandedChange}
                onKeyDown={handleKeyDown}
                className={cn(styles.container, containerClassName)}
            >
                <div
                    className={cn(styles.header, headerClassName, {
                        [styles.endPosition]: isStartPosition,
                    })}
                >
                    {headerContent}
                </div>

                <div
                    className={cn(styles.control, controlClassName, {
                        [styles.startPosition]: isStartPosition,
                    })}
                >
                    {controlContent}
                </div>
            </div>

            <div
                ref={contentRef}
                className={cn(styles.body, bodyClassName, { [styles.expandedBody]: isExpanded })}
            >
                <div className={cn(styles.bodyContent)} ref={contentCaseRef}>
                    {bodyContent}
                </div>
            </div>
        </div>
    );
};
