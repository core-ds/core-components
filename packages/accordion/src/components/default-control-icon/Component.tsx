import React, { FC } from 'react';
import cn from 'classnames';

import { ChevronDownMIcon } from '@alfalab/icons-glyph/ChevronDownMIcon';

import styles from './index.module.css';

type Props = {
    expanded: boolean;
    startPosition: boolean;
};
export const DefaultControlIcon: FC<Props> = ({ expanded, startPosition }) => (
    <ChevronDownMIcon
        className={cn(styles.icon, {
            [styles.expanded]: expanded,
            [styles.startPosition]: startPosition,
        })}
    />
);
