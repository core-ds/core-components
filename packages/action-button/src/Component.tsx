import React, {
    AnchorHTMLAttributes,
    ButtonHTMLAttributes,
    forwardRef,
    Fragment,
    useRef,
} from 'react';
import mergeRefs from 'react-merge-refs';
import cn from 'classnames';

import { Loader } from '@alfalab/core-components-loader';
import { useFocus } from '@alfalab/hooks';

import { useLoader } from './hooks';

import defaultColors from './default.module.css';
import styles from './index.module.css';
import invertedColors from './inverted.module.css';
import staticColors from './static.module.css';

const colorStyles = {
    default: defaultColors,
    inverted: invertedColors,
    static: staticColors,
};

/**
 * Минимальное время отображения лоадера - 500мс,
 * чтобы при быстрых ответах от сервера кнопка не «моргала».
 */
const LOADER_MIN_DISPLAY_INTERVAL = 500;

type Colors = 'default' | 'inverted' | 'static';

type ComponentProps = {
    /**
     * Иконка кнопки
     */
    icon: React.ReactNode;

    /**
     *  Размер кнопки
     */
    size?: 's';

    /**
     * Тип кнопки
     */
    view?: 'primary' | 'secondary';

    /**
     * Дополнительный класс
     */
    className?: string;

    /**
     * Дополнительный класс для обертки иконки
     */
    iconWrapperClassName?: string;

    /**
     * Значение href для ссылки
     */
    href?: string;

    /**
     * Заблокировать кнопку
     */
    disabled?: boolean;

    /**
     * Показать лоадер
     */
    loading?: boolean;

    /**
     * Id компонента для тестов
     */
    dataTestId?: string;

    /**
     * Палитра, в контексте которой используется кнопка
     */
    colors?: Colors;
};

type AnchorProps = ComponentProps & AnchorHTMLAttributes<HTMLAnchorElement>;
type ButtonProps = ComponentProps & ButtonHTMLAttributes<HTMLButtonElement>;
export type ActionButtonProps = Partial<AnchorProps | ButtonProps>;

export const ActionButton = forwardRef<HTMLAnchorElement | HTMLButtonElement, ActionButtonProps>(
    (
        {
            className,
            icon,
            children,
            href,
            size = 's',
            view = 'primary',
            type = 'button',
            iconWrapperClassName,
            disabled,
            loading,
            dataTestId,
            colors = 'default',
            ...rest
        },
        ref,
    ) => {
        const componentRef = useRef<HTMLElement>(null);

        const [focused] = useFocus(componentRef, 'keyboard');

        const { showLoader } = useLoader(!!loading, LOADER_MIN_DISPLAY_INTERVAL);

        const componentProps = {
            className: cn(
                styles.component,
                colorStyles[colors][view],
                styles[size],
                {
                    [styles.focused]: focused,
                    [styles.disabled]: disabled,
                    [styles.loading]: showLoader,
                },
                className,
            ),
            'data-test-id': dataTestId,
        };

        const buttonChildren = (
            <Fragment>
                <span
                    role='img'
                    className={cn(
                        styles.iconWrapper,
                        colorStyles[colors].iconWrapper,
                        styles[size],
                        iconWrapperClassName,
                    )}
                >
                    {showLoader ? <Loader dataTestId='loader' /> : icon}
                </span>
                <span className={styles.label}>{children}</span>
            </Fragment>
        );

        if (href) {
            return (
                <a
                    role='button'
                    ref={mergeRefs([componentRef, ref])}
                    href={href}
                    aria-disabled={disabled || loading}
                    {...componentProps}
                    {...(rest as AnchorHTMLAttributes<HTMLAnchorElement>)}
                >
                    {buttonChildren}
                </a>
            );
        }

        return (
            <button
                ref={mergeRefs([componentRef, ref])}
                // eslint-disable-next-line react/button-has-type
                type={type as ButtonHTMLAttributes<HTMLButtonElement>['type']}
                disabled={disabled || loading}
                {...componentProps}
                {...(rest as ButtonHTMLAttributes<HTMLButtonElement>)}
            >
                {buttonChildren}
            </button>
        );
    },
);
