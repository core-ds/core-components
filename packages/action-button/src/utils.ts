import { getDataTestId } from '@alfalab/core-components-shared';

export function getActionButtonTestIds(dataTestId: string) {
    return {
        actionButton: dataTestId,
        spinner: getDataTestId(dataTestId, 'loader'),
    };
}
