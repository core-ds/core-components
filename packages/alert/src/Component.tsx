import React, { FC } from 'react';
import cn from 'classnames';

import { type Plate, type PlateProps } from '@alfalab/core-components-plate';
import { type PlateDesktop } from '@alfalab/core-components-plate/desktop';
import { type PlateMobile } from '@alfalab/core-components-plate/mobile';

import styles from './index.module.css';

export type AlertProps = Omit<PlateProps, 'foldable' | 'defaultFolded' | 'leftAddons'>;

type AlertPrivateProps = {
    Plate: typeof Plate | typeof PlateDesktop | typeof PlateMobile;
};

export const AlertBase: FC<AlertProps & AlertPrivateProps> = ({
    className,
    title,
    Plate,
    ...restProps
}) => (
    <Plate
        className={cn(styles.component, className)}
        title={title ? <span className={styles.title}>{title}</span> : null}
        rounded={false}
        limitContentWidth={false}
        {...restProps}
    />
);
