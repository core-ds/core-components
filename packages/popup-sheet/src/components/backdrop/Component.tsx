import React, { FC } from 'react';

import { Backdrop, BackdropProps } from '@alfalab/core-components-backdrop';

import styles from './index.module.css';

export type PopupBackdropProps = BackdropProps & {
    /**
     * Прозрачность бэкдропа
     */
    opacity?: number;

    /**
     * Время анимации opacity
     */
    opacityTimeout?: number;
};

export const PopupBackdrop: FC<BackdropProps> = ({
    opacity,
    opacityTimeout,
    style,
    ...backdropProps
}) => (
    <div
        style={{
            opacity,
            transition: opacity === 1 ? `opacity ${opacityTimeout}ms ease-in-out` : '',
            position: 'relative',
            ...style,
        }}
    >
        <Backdrop transitionClassNames={styles} {...backdropProps} />
    </div>
);
