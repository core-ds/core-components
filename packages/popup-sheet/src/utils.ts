import { getDataTestId } from '@alfalab/core-components-shared';

export function getPopupSheetTestIds(dataTestId: string) {
    return {
        popupSheet: dataTestId,
        closer: getDataTestId(dataTestId, 'closer'),
    };
}
