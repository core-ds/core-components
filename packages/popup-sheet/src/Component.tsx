import React, {
    CSSProperties,
    forwardRef,
    KeyboardEvent,
    MouseEvent,
    useRef,
    useState,
} from 'react';
import { SwipeCallback, useSwipeable } from 'react-swipeable';
import cn from 'classnames';

import { BaseModal, BaseModalProps } from '@alfalab/core-components-base-modal';
import { Closer } from '@alfalab/core-components-navigation-bar-private/shared';
import { createPaddingStyle, easingFns, getDataTestId } from '@alfalab/core-components-shared';

import { PaddingType } from '../../types';

import { PopupBackdrop } from './components/backdrop';

import styles from './index.module.css';

const SWIPE_VELOCITY = 0.3;
const CLOSE_OFFSET = 0.3;
const ANIMATION_DURATION = 350;

export type PopupSheetProps = Omit<BaseModalProps, 'onClose' | 'dataTestId'> & {
    /**
     * Наличие кнопки закрытия
     */
    hasCloser?: boolean;

    /**
     * Будет ли свайпаться шторка
     */
    swipeable?: boolean;

    /**
     * Отступы
     */
    padding?: PaddingType;

    /**
     * Обработчик закрытия
     */
    onClose?: (
        event: MouseEvent<HTMLElement> | KeyboardEvent<HTMLElement>,
        reason?: 'backdropClick' | 'escapeKeyDown' | 'closerClick' | 'swipe',
    ) => void;

    /**
     * Идентификатор для систем автоматизированного тестирования.
     * Для кнопки закрытия используется модификатор -closer
     */
    dataTestId?: string;
};

const DEFAULT_PADDING = 32;

export const PopupSheet = forwardRef<HTMLDivElement, PopupSheetProps>(
    (
        {
            children,
            hasCloser,
            className,
            onClose,
            transitionProps,
            contentProps,
            swipeable,
            dataTestId,
            backdropProps,
            padding = DEFAULT_PADDING,
            ...restProps
        },
        ref,
    ) => {
        const [sheetOffset, setSheetOffset] = useState(0);
        const [backdropOpacity, setBackdropOpacity] = useState(1);

        const sheetHeight = useRef(0);
        const sheetRef = useRef<HTMLDivElement>(null);
        const bySwipeCloseAnimation = useRef(false);

        const getSheetOffset = (deltaY: number): number => Math.max(0, deltaY);

        const getBackdropOpacity = (offset: number): number =>
            Math.max(0, 1 - offset / sheetHeight.current);

        const getSwipeStyles = (): CSSProperties =>
            sheetOffset ? { transform: `translateY(${sheetOffset}px)` } : {};

        const handleSwiping: SwipeCallback = ({ deltaY }) => {
            const offset = getSheetOffset(deltaY);

            setSheetOffset(offset);
            setBackdropOpacity(getBackdropOpacity(offset));
        };

        const handleSwiped: SwipeCallback = ({ deltaY, velocity, event }) => {
            const shouldCloseByVelocity = deltaY >= 0 && velocity >= SWIPE_VELOCITY;
            const shouldCloseByOffset = deltaY >= sheetHeight.current * CLOSE_OFFSET;

            const offset = sheetOffset;
            const animDuration = ANIMATION_DURATION / 2;
            let start: number;

            if (shouldCloseByVelocity || shouldCloseByOffset) {
                bySwipeCloseAnimation.current = true;
                onClose?.(event as MouseEvent<HTMLElement>, 'swipe');

                const runLoop = (timeStamp: number) => {
                    if (!start) start = timeStamp;
                    const elapsedTime = timeStamp - start;

                    if (elapsedTime <= animDuration) {
                        const nextOffset =
                            offset +
                            (sheetHeight.current - offset) *
                                easingFns.easeInOutQuad(elapsedTime / animDuration);

                        if (nextOffset > 0) {
                            setSheetOffset(nextOffset);
                        }
                        requestAnimationFrame(runLoop);
                    } else {
                        setSheetOffset(0);
                        bySwipeCloseAnimation.current = false;
                    }
                };

                requestAnimationFrame(runLoop);
            } else {
                const runLoop = (timeStamp: number) => {
                    if (!start) start = timeStamp;
                    const elapsedTime = timeStamp - start;

                    if (elapsedTime <= animDuration) {
                        const nextOffset =
                            offset - offset * easingFns.easeInOutQuad(elapsedTime / animDuration);

                        if (nextOffset > 0) {
                            setSheetOffset(nextOffset);
                            setBackdropOpacity(getBackdropOpacity(nextOffset));
                        }
                        requestAnimationFrame(runLoop);
                    } else {
                        setSheetOffset(0);
                        setBackdropOpacity(1);
                    }
                };

                requestAnimationFrame(runLoop);
            }
        };

        const handleEntered = (node: HTMLElement, isAppearing: boolean) => {
            bySwipeCloseAnimation.current = false;
            if (sheetRef.current) {
                sheetHeight.current = sheetRef.current.offsetHeight;
            }

            if (transitionProps?.onEntered) {
                transitionProps.onEntered(node, isAppearing);
            }
        };

        const handleExited = (node: HTMLElement) => {
            bySwipeCloseAnimation.current = false;
            setSheetOffset(0);
            setBackdropOpacity(1);

            if (transitionProps?.onExited) {
                transitionProps.onExited(node);
            }
        };

        const sheetSwipeableHandlers = useSwipeable({
            onSwiping: handleSwiping,
            onSwiped: handleSwiped,
            trackMouse: swipeable,
            trackTouch: swipeable,
            delta: 5,
        });

        return (
            <BaseModal
                {...restProps}
                onClose={onClose}
                ref={ref}
                className={cn(styles.component, className)}
                dataTestId={dataTestId}
                Backdrop={PopupBackdrop}
                backdropProps={{
                    ...backdropProps,
                    opacity: backdropOpacity,
                }}
                transitionProps={{
                    classNames: {
                        enter: styles.enter,
                        appear: styles.appear,
                        enterActive: styles.enterActive,
                        appearActive: styles.appearActive,
                        exit: bySwipeCloseAnimation.current ? styles.exitBySwipe : styles.exit,
                        exitActive: bySwipeCloseAnimation.current
                            ? styles.exitActiveBySwipe
                            : styles.exitActive,
                    },
                    timeout: ANIMATION_DURATION,
                    ...transitionProps,
                    onEntered: handleEntered,
                    onExited: handleExited,
                }}
                componentDivProps={{
                    ref: sheetRef,
                    style: getSwipeStyles(),
                }}
                contentProps={{
                    style: createPaddingStyle(padding),
                    ...contentProps,
                    ...sheetSwipeableHandlers,
                    className: cn(styles.content, contentProps?.className),
                }}
            >
                {hasCloser && (
                    <Closer
                        view='mobile'
                        className={styles.closer}
                        onClick={onClose}
                        dataTestId={getDataTestId(dataTestId, 'closer')}
                    />
                )}
                {children}
            </BaseModal>
        );
    },
);
