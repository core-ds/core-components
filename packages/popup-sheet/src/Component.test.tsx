import React, { ReactNode } from 'react';
import { PopupSheet, PopupSheetProps } from './Component';
import { act, render, RenderResult } from '@testing-library/react';
import { getPopupSheetTestIds } from './utils';

const renderPopupSheet = async (
    props: PopupSheetProps & { children: ReactNode },
): Promise<RenderResult | undefined> => {
    let result: RenderResult | undefined;

    await act(async () => {
        result = await render(<PopupSheet {...props}>{props.children}</PopupSheet>);
    });

    return result;
};

describe('PopupSheet', () => {
    Object.defineProperty(window, 'matchMedia', {
        writable: true,
        value: jest.fn().mockImplementation((query) => ({
            matches: true,
            media: query,
            onchange: null,
            addListener: jest.fn(), // Deprecated
            removeListener: jest.fn(), // Deprecated
            addEventListener: jest.fn(),
            removeEventListener: jest.fn(),
            dispatchEvent: jest.fn(),
        })),
    });
    describe('Snapshot tests', () => {
        it('should match snapshot', async () => {
            const component = await renderPopupSheet({
                children: <div style={{ height: 50 }}>PopupSheet</div>,
                open: true,
            });

            expect(component).toMatchSnapshot();
        });
    });

    describe('attributes test', () => {
        it('should have data-test-id', () => {
            const dti = 'popup-sheet-dti';

            const { getByTestId } = render(
                <PopupSheet hasCloser={true} open={true} dataTestId={dti}>
                    <div style={{ height: 50 }} />
                </PopupSheet>,
            );

            const testIds = getPopupSheetTestIds(dti);

            expect(getByTestId(testIds.popupSheet)).toBeInTheDocument();
            expect(getByTestId(testIds.closer)).toBeInTheDocument();
        });
    });
});
