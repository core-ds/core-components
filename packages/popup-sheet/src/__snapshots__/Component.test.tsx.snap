// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PopupSheet Snapshot tests should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    style="padding-right: 0px; overflow: hidden;"
  >
    <div />
    <div
      alfa-portal-container=""
    >
      <div
        data-focus-guard="true"
        style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
        tabindex="0"
      />
      <div
        data-focus-lock-disabled="false"
      >
        <div
          style="opacity: 1; transition: opacity undefinedms ease-in-out; position: relative; z-index: 100;"
        >
          <div
            aria-hidden="true"
            class="backdrop appear appearActive"
          />
        </div>
        <div
          class="wrapper"
          role="dialog"
          style="z-index: 100;"
          tabindex="-1"
        >
          <div
            class="component appear appearActive"
          >
            <div
              class="content content"
              style="padding: 32px;"
            >
              <div
                style="height: 50px;"
              >
                PopupSheet
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        data-focus-guard="true"
        style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
        tabindex="0"
      />
    </div>
  </body>,
  "container": <div />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
